### PR TITLE
[TextFields] Prevents functionally equivalent fonts from causing rendering jumps

### LINKED ISSUE #4189: [Textfield] Floating Placeholder text changes size when you type in the 2nd word into the text field
In a controller make a MDCTextfield a wrap a MDCTextInputControllerFilled controller on it

let textField:BITextField = BITextField()
MDCTextInputControllerFilled(textInput: textField)

In the text field type 1 letter in, the floating placeholder text will get smaller and go on top

When you type in a 2nd letter in, the floating placeholder text will get smaller again.

Expected behavior: The floating placeholder text should remain the same size and not change

---
#### Internal data
- Associated internal bug: [b/117178829](http://b/117178829)

### PR DESCRIPTION
**Pull request by willlarche**
Helps with but doesn't close #4189

During the `updatePlaceholder`, fonts were previously being applied without checking to see if they were even different and therefore warranted a change. This adds that missing comparison.

But comparing fonts is hard. If you have a font with a `UIFontDescriptor` that has family `.UISFText`, weight 'regular', and size `16` it will not return `YES` when compared with a font that is the same family, size, and weight but also a text style like `Body`. It will resolve to the same font as the other one but Apple will render it 1 pixel wider or narrower because that symbolic trait is missing. Confirmed with our Fonts team that this is a quirk of Apple's text rendering. So, to prevent that little jump, I've added a custom check for all the parts of a font that matter.

### CODE REVIEW

File: components/Typography/src/UIFont+SimpleEquality.m
Comment (jverkoey) [inline components/Typography/src/UIFont+SimpleEquality.m]: minor style: please wrap these conditionals on separate lines.

Comment (willlarche) [reply components/Typography/src/UIFont+SimpleEquality.m]: Sorry! Rusty. Ran format on everything I did.

File: components/Typography/src/UIFont+SimpleEquality.m
Comment (jverkoey) [inline components/Typography/src/UIFont+SimpleEquality.m]: I think that UIFontDescriptorFaceAttribute returns a string, so we may need to use `isEqual:` here.

File: components/Typography/tests/unit/FontEqualityTests.m
Comment (jverkoey) [inline components/Typography/tests/unit/FontEqualityTests.m]: minor: Copyright 2018

File: components/Typography/src/UIFont+SimpleEquality.h
Comment (jverkoey) [inline components/Typography/src/UIFont+SimpleEquality.h]: Because this is an extension it will require an mdc_ prefix. Consider adding a qualifier to this as well to indicate that we're doing deeper equality checking than the default isEqual: implementation.

---

Alternatively — because it's possible we might re-implement this same extension elsewhere — it may be worth considering making this a C API. This will de-risk the possibility of accidentally introducing this extension in multiple places within MDC in the future and having to deal with very difficult-to-debug behavior.

For context, duplicate Objective-C extension APIs will build successfully with no warning (though if this has changed recently that would be wonderful), resulting in one of the implementations being chosen (essentially at random) at runtime. C APIs on the other hand will not pass the linker phase if multiple APIs exist with the same symbol.

So the end result would look roughly like so:

```objc
FOUNDATION_EXPORT
BOOL MDCFontIsDeeplyEqualToFont(UIFont *font1, UIFont *font2);
```

**Review by jverkoey — CHANGES_REQUESTED** [review]

File: components/Typography/src/UIFont+SimpleEquality.m
Comment (jverkoey) [inline components/Typography/src/UIFont+SimpleEquality.m]: We ran into similar comparison fun in #3360 related to `UIFontDescriptorTraitsAttribute` being different. It may be worth checking that the `UIFontDescriptorTraitsAttribute` matches up on each font as well.

File: components/TextFields/src/MDCTextInputControllerBase.m
Comment (mohammadcazig) [inline components/TextFields/src/MDCTextInputControllerBase.m]: Can you leave a comment here explaining why we have to do this?

File: components/Typography/src/UIFont+SimpleEquality.m
Comment (willlarche) [reply components/Typography/src/UIFont+SimpleEquality.m]: Done

File: components/Typography/tests/unit/FontEqualityTests.m
Comment (willlarche) [reply components/Typography/tests/unit/FontEqualityTests.m]: Done

File: components/Typography/src/UIFont+SimpleEquality.h
Comment (willlarche) [reply components/Typography/src/UIFont+SimpleEquality.h]: Done.

File: components/Typography/src/UIFont+SimpleEquality.m
Comment (willlarche) [reply components/Typography/src/UIFont+SimpleEquality.m]: Done.

File: components/TextFields/src/MDCTextInputControllerBase.m
Comment (willlarche) [reply components/TextFields/src/MDCTextInputControllerBase.m]: Done.

**Comment by willlarche** [comment]
@jverkoey PTAL

File: components/Typography/src/UIFont+MaterialSimpleEquality.h
Comment (jverkoey) [inline components/Typography/src/UIFont+MaterialSimpleEquality.h]: You don't need the `@interface` surrounding the C method.

File: components/Typography/src/UIFont+MaterialSimpleEquality.h
Comment (jverkoey) [inline components/Typography/src/UIFont+MaterialSimpleEquality.h]: Docs look like they need to be updated:

> UIFont's `isEqual:` is more 'shallow' than this method's implementation.

File: components/Typography/src/UIFont+MaterialSimpleEquality.m
Comment (jverkoey) [inline components/Typography/src/UIFont+MaterialSimpleEquality.m]: Is this comment relevant here?

**Review by jverkoey — APPROVED** [review]
LGTM pending review feedback being addressed.

File: components/Typography/src/UIFont+MaterialSimpleEquality.m
Comment (willlarche) [reply components/Typography/src/UIFont+MaterialSimpleEquality.m]: Nope. Over-copy. Thank you.

File: components/Typography/src/UIFont+MaterialSimpleEquality.h
Comment (willlarche) [reply components/Typography/src/UIFont+MaterialSimpleEquality.h]: No, it's correct. Mine only checks 4 properties. I think `isEqual:` checks more.

Comment (jverkoey) [reply components/Typography/src/UIFont+MaterialSimpleEquality.h]: Ah sorry, I should have been more clear that the “isEqualToFont:” reference is, I believe, no longer applicable.

Comment (jverkoey) [reply components/Typography/src/UIFont+MaterialSimpleEquality.h]: Ah sorry, I should have been more clear that the “isEqualToFont:” reference is, I believe, no longer applicable.

File: components/Typography/src/UIFont+MaterialSimpleEquality.h
Comment (willlarche) [reply components/Typography/src/UIFont+MaterialSimpleEquality.h]: Done.

File: components/Typography/src/UIFont+MaterialSimpleEquality.h
Comment (willlarche) [reply components/Typography/src/UIFont+MaterialSimpleEquality.h]: Oh oh oh. Sorry. yes. Thank you. Done!

**Comment by jverkoey** [comment]
This PR is blocked on CI.

```

Test Case '-[FontEqualityTests testSimpleFontEquality]' started.
components/Typography/tests/unit/FontEqualityTests.m:36: error: -[FontEqualityTests testSimpleFontEquality] : ((MDCFontIsSimplyEqualToFont(baseFont, additionalAttributeFont)) is true) failed
Test Case '-[FontEqualityTests testSimpleFontEquality]' failed (0.032 seconds).
Test Suite 'FontEqualityTests' failed at 2018-06-01 18:14:10.340.
	 Executed 1 test, with 1 failure (0 unexpected) in 0.032 (0.032) seconds
```

**Comment by jverkoey** [comment]
Looks like the fonts are being loaded correctly, so you may want to add an NSLog to the test dumping the values being compared in order to see what's different.

**Comment by willlarche** [comment]
Issue is the traits attribute:

```
{
    NSCTFontSlantTrait = 0;
    NSCTFontSymbolicTrait = 16384;
    NSCTFontUIFontDesignTrait = NSCTFontUIFontDesignDefault;
    NSCTFontWeightTrait = 0;
}

{
    NSCTFontSlantTrait = 0;
    NSCTFontSymbolicTrait = 16384;
    NSCTFontWeightTrait = 0;
}
```

@jverkoey Mind if I take a check for that out? I think it's the one you had recommended.

**Comment by jverkoey** [comment]
These traits will affect the kerning of the font rendering, so if you are attempting to do a visual equality check then that comparison likely has to stay.

**Comment by willlarche** [comment]
Removed that check after all. merging.

**Comment by yarneo** [comment]
@willlarche unfortunately this PR had to be rolled back. I opened an issue for tracking: https://github.com/material-components/material-components-ios/issues/4385

**Comment by willlarche** [comment]
@yarneo I already deleted the branch on my fork. What do I do to update this?

**Pull request merged**
